### ISSUE TITLE
MS map centering not correct

### ISSUE THREAD

**Issue by tdipisa** [geosolutions-it/MapStore2#6732]
## Description

We have the following user feedback:

>When searching for a point feature it is not possible to change the zoom level, without losing the feature out of focus. 

## How to reproduce

- Search for LOAG on https://austrocontrol.geo-solutions.it/austrocontrol-ms2/#/viewer/openlayers/763 (using the search tool)
- Try to zoom in by using a different scale from the dropdown list (scale selector)

![zoom_on_search](https://user-images.githubusercontent.com/1280027/114056939-7590c080-9892-11eb-807b-a9ba26fe0239.gif)

*Expected Result*

 Search results should always be centered so the user can also use the drop down zoom change.

*Current Result*

Zooming using the scale selector we loose the search result in map: the zoom operation is not centered on the marker.

- [x] Not browser related

<details><summary> <b>Browser info</b> </summary>

(use this site: <a href="https://www.whatsmybrowser.org/">https://www.whatsmybrowser.org/</a> for non expert users)

| Browser Affected | Version |
|---|---|
|Internet Explorer| |
|Edge| |
|Chrome| |
|Firefox| |
|Safari| |
</details>

**Comment by tdipisa** [comment]
@offtherailz any thought on this?

**Comment by offtherailz** [comment]
I think there is a little misalignment of a 30 pixels from the bottom when zoom to extent. (due to maplayout).
https://github.com/geosolutions-it/MapStore2/blob/f86ddf8916d134154c4cf395cf7d790da9e1e7fe/web/client/epics/maplayout.js#L80

This on every zoom level the error doubles (30, 60, 120, 240, 580).

That space has been added (correct me if I'm wrong @allyoucanmap ) to leave space for the attribution, when the footer is not present. 

The issue can be fixed by overriding that value setting in `localConfig.json` (or modifying the current default):
```
 "mapLayout":
      {
      "left": {"sm": 300, "md": 500, "lg": 600},
      "right": {"md": 658},
      "bottom": {"sm": 0}
    },
```
_note I changed bottom  --> sm: 30 to 0_

Having the issue fixed and this graphical change:
before: 
![image](https://user-images.githubusercontent.com/1279510/113326054-7e751580-9319-11eb-9fe7-510bd0a9815b.png)
after:
![image](https://user-images.githubusercontent.com/1279510/113326043-7b7a2500-9319-11eb-8c42-fce28dafd39a.png)

I'm wandering if makes sense to keep it @allyoucanmap (or we can set it to 0 as default)

**Comment by allyoucanmap** [comment]
@offtherailz the 30px bottom is not realted to the map visibility but the general layout. If you change that value from 30 to 0 the footer is still visible when the attribute table is open?

**Comment by offtherailz** [comment]
It is visible, I see only this:
![image](https://user-images.githubusercontent.com/1279510/114055895-707f4180-9891-11eb-8d33-142768f168ad.png)

That doesn't appen in mapstore, but maybe is a bug in the implementation.

**Comment by allyoucanmap** [comment]
@offtherailz @MuhweziDeo 
the [maplayout reducer](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/reducers/maplayout.js#L17) stores two properties: layout and boundingMapRect.
The layout one is used specifically to align the plugins on the map while the boundingMapRect defines the visible map in the app. 
The zoomToExtent action triggers an epic that use of the [mapPaddingSelector](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/epics/map.js#L194) selector to provide the padding option. The mapPaddingSelector uses the [boundingMapRect](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/selectors/maplayout.js#L102-L111) property. So probably we should change this part https://github.com/geosolutions-it/MapStore2/blob/master/web/client/epics/maplayout.js#L120-L124 and ensure the bottom is set to 0 instead to use the default value only in case of the attribute table is closed. Semething like this (I did not test this):
```js
const mapBottom = isFeatureGridOpen(state) && {bottom: dockSize + '%', dockSize} || { };
const boundingMapRect = {
  ...mapBottom,
  ...leftPanels,
  ...rightPanels
};

```

**Comment by MuhweziDeo** [comment]
> @offtherailz @MuhweziDeo
> the [maplayout reducer](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/reducers/maplayout.js#L17) stores two properties: layout and boundingMapRect.
> The layout one is used specifically to align the plugins on the map while the boundingMapRect defines the visible map in the app.
> The zoomToExtent action triggers an epic that use of the [mapPaddingSelector](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/epics/map.js#L194) selector to provide the padding option. The mapPaddingSelector uses the [boundingMapRect](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/selectors/maplayout.js#L102-L111) property. So probably we should change this part https://github.com/geosolutions-it/MapStore2/blob/master/web/client/epics/maplayout.js#L120-L124 and ensure the bottom is set to 0 instead to use the default value only in case of the attribute table is closed. Semething like this (I did not test this):
> 
> ```js
> const mapBottom = isFeatureGridOpen(state) && {bottom: dockSize + '%', dockSize} || { };
> const boundingMapRect = {
>   ...mapBottom,
>   ...leftPanels,
>   ...rightPanels
> };
> ```
let me test this out on my end and see but had realized that the spacing was coming from this class `mapToolbar` and this `background-plugin-position` but I guess you are right setting bottom to `0` brought about this change

**Comment by MuhweziDeo** [comment]
@allyoucanmap I have tested your solution above and seems not to work. Maybe am wrong but I think the issue is mainly because of the CSS classes eg `background-plug-position` https://github.com/geosolutions-it/MapStore2/blob/master/web/client/themes/default/less/backgroundSelector.less#L7 which explicit values for the margin-bottom and I think this worked fine when the bottom was `30`. I have worked on solution that makes the margin-bottom value dynamic not sure if it will cause issues else but you can test and let me know

**Comment by MuhweziDeo** [comment]
> this

hey @allyoucanmap 
Redid a test using this seems to work but there is a misalignment on the part of the `background-selector`

**Comment by allyoucanmap** [comment]
from a second investigation the bottom 30px could not be the only problem related to this issue. In the scheme below you ca see in red the square representing half map while in black the square representing the size of 30px x 30px. I think we should ensure the following:

- if the location of the anchor point of the marker is at the base or at the center
- if there is an error of precision due to the increasing zoom level

![image](https://user-images.githubusercontent.com/19175505/114185022-864e3e80-9945-11eb-9bdd-07fc33786675.png)

We could start to try:
- option 1 use a specific bottom params for the boundingMapRect
- option 2 restore the bottom: 0 in the css of the #map class to make the layout behaviour consistent with the map (the CSS was introduced here and maybe could be restored to 0 now https://github.com/geosolutions-it/MapStore2/commit/3f1dc4d1540283888dd96152ebb8c0dcb2a2d610)

if both are not solving the issue and we continuing have the misalignement due to the scale precision we should discuss how to proceed (cc @offtherailz )

**Comment by MuhweziDeo** [comment]
Okay, I have tried both the recommended solutions but none of them seem to fix the issue. But I discovered a pattern. 

1. **Restoring CSS for the bottom to 0**
With this as the zoom increases, the marker continues to disappear but it happens from the top, and also there is space left below the map I think as a result of removing the `bottom: 28px`

<img width="1440" alt="Screenshot 2021-04-09 at 16 36 09" src="https://user-images.githubusercontent.com/39124174/114189314-be0fb300-9952-11eb-9fc4-bfac2bcb43f7.png">

2. Setting the value for the bottom to zero in `epics/maplayout.js`

This also does not fix the issue but the marker disappears from the bottom the more the zoom increases.
<img width="1440" alt="Screenshot 2021-04-09 at 16 34 18" src="https://user-images.githubusercontent.com/39124174/114189554-fd3e0400-9952-11eb-9728-650837c8e80e.png">

cc @offtherailz @allyoucanmap

**Comment by dsuren1** [comment]
As per discussion with @offtherailz and @allyoucanmap , 

**Possible issue**: The zoom to extent uses padding (see https://openlayers.org/en/latest/apidoc/module-ol_View-View.html#fit, leaflet has a similar system) and this de-center the desired center. 
The zoom buttons or the zoom selector only change the zoom level in the state, without re-centering anything, so this looks to be the source of the problem. Also the style in mapLayout may influence it as well. 

**Investigation**: So we tried with setting bottom value as 0 only to boundingMapRect while maintaining the layout style untouched.

**Result**: 
1. The layout bottom allows to fit the components on the screen, while boundingMapRect is the map padding
2. Also it eliminates the map decentering from zoom level

**Comment by tdipisa** [comment]
@offtherailz that's right. @dsuren1 please keep the MS revision of the c125 project upated so that we can test in c125 dev before porting this fix in master and 2021.01.xx

**Comment by tdipisa** [comment]
@ElenaGallo I saw you have already tested this in c125 project. Do you still have to test it in MS DEV?
